pkg/gcpwatch/tracker.go

```diff
@@ -0,0 +1,237 @@
+
+		if err := t.pollOnce(ctx); err != nil {
+			klog.Warningf("error during drift-correction polling: %v", err)
+		}
+	}
+}
+
+func (t *DependencyTracker) pollOnce(ctx context.Context) error {
+	log := klog.FromContext(ctx)
+
+	var keys []gcpResouceKey
+	t.mutex.Lock()
+	for key := range t.gcpResources {
+		keys = append(keys, key)
+	}
+	t.mutex.Unlock()
+
+	for _, key := range keys {
+		t.mutex.Lock()
+		gcpResource := t.gcpResources[key]
+		t.mutex.Unlock()
+
+		if gcpResource == nil {
+			continue
+		}
+
+		latest, err := t.fetcher.Fetch(ctx, key.Kind, key.External)
+		if err != nil {
+			// TODO: Remove if not found?
+			log.Error(err, "error fetching iam", "kind", key.Kind, "external", key.External)
+			continue
+		}
+
+		gcpResource.mutex.Lock()
+		log.Info("got iam etag", "newEtag", latest.Etag, "kind", key.Kind, "external", key.External, "oldEtag", gcpResource.etag)
+		if latest.Etag != gcpResource.etag {
+			gcpResource.etag = latest.Etag
+
+			for _, dep := range gcpResource.dependencies {
+				log.Info("triggering notification", "controller", dep.controller.controllerKey, "namespace", dep.namespace, "name", dep.name)
+				genEvent := event.GenericEvent{}
+				genEvent.Object = &unstructured.Unstructured{}
+				genEvent.Object.SetNamespace(dep.namespace)
+				genEvent.Object.SetName(dep.name)
+				dep.controller.queue <- genEvent
+			}
+		}
+		gcpResource.mutex.Unlock()
+	}
+
+	return nil
+}
+
+func (r *ControllerRegistration) Add(ctx context.Context, policy *v1beta1.IAMPartialPolicy, etag string) bool {
+	log := klog.FromContext(ctx)
+
+	if etag == "" {
+		return false
+	}
+
+	if r == nil {
+		return false
+	}
+
+	// TODO: Resolve Name -> External
+
+	t := r.tracker
+	kind := policy.Spec.ResourceReference.Kind
+	if kind == "" {
+		return false
+	}
+
+	external := policy.Spec.ResourceReference.External
+	if external == "" {
+		return false
+	}
+
+	if !t.fetcher.IsSupported(kind, external) {
+		return false
+	}
+
+	log.Info("adding watch on policy", "policy", policy, "kind", kind, "external", external)
+
+	t.add(kind, external, r, policy.GetNamespace(), policy.GetName(), etag)
+	return true
+}
+
+func (t *DependencyTracker) add(kind string, external string, controller *ControllerRegistration, namespace string, name string, etag string) {
+	target := t.getTarget(kind, external)
+
+	target.mutex.Lock()
+	defer target.mutex.Unlock()
+
+	if target.etag == "" {
+		target.etag = etag
+		// TODO: Trigger now?
+	}
+
+	for i := range target.dependencies {
+		dep := &target.dependencies[i]
+		if dep.controller == controller && dep.namespace == namespace && dep.name == name {
+			return
+		}
+	}
+	target.dependencies = append(target.dependencies, dependency{
+		controller: controller,
+		namespace:  namespace,
+		name:       name,
+	})
+}
+
+func (t *DependencyTracker) getTarget(kind string, external string) *dependenciesByResource {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	key := gcpResouceKey{
+		Kind:     kind,
+		External: external,
+	}
```

Comment:
In current KCC, what happens when a reference changes from external to KNN? i.e. a ProjectRef used to not have CC Project object, and a LLM use ProjectRef.external. then it has a CC object. Will the LLM controller update the projectRef.name and projectRef.namespace with corresponding value? If so, if the CC project is changed, will that trigger the LLM object to reconcile?